src/zypper_plugins/tracer_upload.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/python
+#
+# Copyright 2019 ATIX AG
+#
+# This software is licensed to you under the GNU General Public
+# License as published by the Free Software Foundation; either version
+# 2 of the License (GPLv2) or (at your option) any later version.
+# There is NO WARRANTY for this software, express or implied,
+# including the implied warranties of MERCHANTABILITY,
+# NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
+
+import json
+import httplib
+from os import path, environ
+import sys
+import logging
+from rhsm.config import RhsmConfigParser, initConfig
+from zypp_plugin import Plugin
+
+sys.path.append('/usr/share/rhsm')
+from subscription_manager.identity import ConsumerIdentity
+
+REBOOT_NEEDED_FLAG = "/var/run/reboot-needed"
+
+class TracerUploadPlugin(Plugin):
+    def __init__(self):
+        Plugin.__init__(self)
+        self.num1 = self.num2 = None
+        self.description = "Upload traces to katello"
+        self.cleanup = ""
+        self.userdat = {}
+
+    def collect_data(self):
+        """
+        Return a array with nested arrays
+        containing name, how to restart & app type
+        for every package that needs restarting
+        """
+
+        apps = {}
+
+        if path.isfile(REBOOT_NEEDED_FLAG)
+            apps["kernel"] = { "helper": "You will have to reboot your computer", "type": "static" }
+        return apps
+
+    def upload_tracer_profile(self):
+        data =  json.dumps({ "traces": self.collect_data() })
+        headers = { "Content-Type": "application/json" }
+
+        conn = httplib.HTTPSConnection(
+            RhsmConfigParser.get(initConfig() ,'server', 'hostname'),
+            RhsmConfigParser.get(initConfig() ,'server', 'port'),
+            key_file = ConsumerIdentity.keypath(),
+            cert_file = ConsumerIdentity.certpath()
+        )
+        conn.request('PUT', '/rhsm/consumers/%s/tracer' % (ConsumerIdentity.read().getConsumerId()), data, headers=headers)
+        response = conn.getresponse()
+
+    def PLUGINBEGIN(self, headers, body):
+        logging.info("PLUGINBEGIN")
+        self.ack()
+
+    def COMMITBEGIN(self, headers, body):
+        logging.info("COMMITBEGIN")
+        self.ack()
+
+    def COMMITEND(self, headers, body):
+        logging.info("COMMITEND")
+        self.ack()
+
+    def PLUGINEND(self, headers, body):
+        logging.info("PLUGINEND")
+
+        logging.info("Uploading Tracer Profile")
+
+        try:
+            self.upload_tracer_profile()
+        except:
+            logging.error("Unable to upload Tracer Profile")
+
+        self.ack()
+
+if "DISABLE_TRACER_UPLOAD_PLUGIN" in environ:
```

Comment:
Python convention is to wrap this behind `if __name__ == '__main__':` so you can import from it if needed.

Reply:
This is used for enabling / disabling the plugin. its used on other zypper plugins, too.

Reply:
I meant
```python
if __name__ == '__main__':
    if "DISABLE_TRACER_UPLOAD_PLUGIN" in environ:
        logging.info("$DISABLE_TRACER_UPLOAD_PLUGIN is set - disabling katello tracer upload plugin")
        # a dummy Plugin is needed
        plugin = Plugin()
    else:
        plugin = TracerUploadPlugin()

    plugin.main()
```